src/Functions/stringToMap.cpp

```diff
@@ -0,0 +1,278 @@
+#include <DataTypes/DataTypeMap.h>
+#include <DataTypes/DataTypeString.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/IFunction.h>
+#include <iostream>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int ILLEGAL_COLUMN;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+    extern const int BAD_ARGUMENTS;
+}
+
+class FunctionStringToMap : public IFunction
+{
+public:
+    static constexpr auto name = "stringToMap";
+    static FunctionPtr create(ContextPtr) { return std::make_shared<FunctionStringToMap>(); }
+
+    String getName() const override
+    {
+        return name;
+    }
+
+    bool isVariadic() const override { return true; }
+    size_t getNumberOfArguments() const override { return 0; }
+
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        size_t num_arguments = arguments.size();
+
+        if (num_arguments < 1 || num_arguments > 3)
+            throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Number of arguments for function {} doesn't match: "
+                            "passed {}, should be 1, 2 or 3", getName(), num_arguments);
+        if (!isStringOrFixedString(arguments[0]))
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of argument of function {}",
+                            arguments[0]->getName(), getName());
+        if (num_arguments >= 2 && !isStringOrFixedString(arguments[1]))
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of argument of function {}",
+                            arguments[1]->getName(), getName());
+        if (num_arguments == 3 && !isStringOrFixedString(arguments[2]))
+            throw Exception(ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT, "Illegal type {} of argument of function {}",
+                            arguments[2]->getName(), getName());
+
+        DataTypePtr string_type = std::make_shared<DataTypeString>();
+        DataTypePtr string_type_nullable = std::make_shared<DataTypeNullable>(string_type);
+        DataTypes key_value_type{string_type, string_type_nullable};
+
+        return std::make_shared<DataTypeMap>(key_value_type);
+    }
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    static bool tryExecuteString(const IColumn * col, ColumnPtr & col_res, char pair_delim, char key_value_delim)
+    {
+        const ColumnString * col_str_in = checkAndGetColumn<ColumnString>(col);
+
+        if (col_str_in)
+        {
+            const ColumnString::Chars & in_vec = col_str_in->getChars();
+            const ColumnString::Offsets & in_offsets = col_str_in->getOffsets();
+
+            size_t size = in_offsets.size();
+
+            MutableColumnPtr keys_data = DataTypeString().createColumn();
```

Comment:
I think we are able to do a `reserve()` or (better) `resize()` on the output structures here based on the size of the input data.

Reply:
- `IColumn` only has a `reserve()` method.
- I can use `reserve()` for the offsets column because the number of offsets will be equal to the number of input rows. So, I can reserve `input_rows_count` size for this column.
- But for the keys and values, I do not know the exact size that the columns are going to need as that would require processing the actual input strings. Although I think we can say  that they are at least going to be the size equal to the the number of rows, i.e., at least one key-value pair per row. So should I reserve `input_rows_count` size for these columns as well ?